lib/redis_client/cluster/node.rb

```diff
@@ -78,14 +79,25 @@ def []=(index, element)
         end
       end
 
+      class SingleNodeRedisClient < ::RedisClient
+        include Pinning::ClientMixin
```

Comment:
I think the following complexity as you said is derived from this mixin pattern.

* Config instance holds the command information.
* Our cluster client uses our own client implementation.
* Pinning module holds the state of the locking.

I think it would be better to use a simple wrapper object for the slot validation.

```ruby
class RedisClient
  class Cluster
    class PinningClient
      def initialize(client, command, command_builder, slot)
        @client = client
        @command = command
        @command_builder = command_builder
        @slot = slot
      end

      def call(*args, **kwargs, &block)
        command = @command_builder.generate(args, kwargs)
        call_v(command, &block)
      end

      def call_v(command, &block)
        validate_slot!(command)
        @client.call_v(command, &block)
      end

      # etc...

      private

      def validate_slot!(command)
        # raise an error if command is invalid
      end
    end
  end
end
```

```ruby
def with(key: nil, hashtag: nil, write: true, retry_count: 0)
  key = process_with_arguments(key, hashtag)

  node_key = @router.find_node_key_by_key(key, primary: write)
  node = @router.find_node(node_key)
  @router.try_delegate(node, :with, retry_count: retry_count) do |conn|
    yield PinningClient.new(conn, @command_builder, @router.command)
  end
end
```

Reply:
We can do this, but it's a little awkward because of the need to re-wrap the client if `pipelined` or `multi` is called. It ends up looking like this: https://github.com/redis-rb/redis-cluster-client/pull/298/files#diff-426e610bbc89f748ae5f9d43790b07d10fed426fa431d6bea29b67f649a7eb72R1 which i remember we were not very happy with.

Most of the complexity here is coming from the fact that we have to change the lifecycle of the `RedisClient::Cluster::Command` object so that it's available for the pinning middleware. If we made our custom `RedisClient` subclass support some kind of temporary "validation procs" itself, then the `RedisCluster::Client` instance can construct the validation proc only when `#with` is called and `@commands` is already known.

I think that will be a lot simpler - let me try that first.

Reply:
I think we can omit `#pipelined` and `#multi` in `#with` method. There is no user to use in the odd way. We can't never care about all use cases completely. Our users should have responsible to use it method.

Reply:
We definitely need `#multi` at least - the whole point is to enable single shard transactions!

I managed to write a much better implementation of the delegator in #298  today though - expect to see another PR tomorrow!

Reply:
I thought the `#multi` methods is holded by the cluster client.

```ruby
@cluster_client = build_cluster_client

@cluster_client.with(...) do |raw_or_pinning_cli|
  raw_or_pinning_cli.call('CMD')

  @cluster_client.multi(...) do |raw_or_pinning_cli|
    # do something
  end
end
```

Anyway, I think we shouldn't do overengineering.

Reply:
No, my plan was to delete the `#multi` implementation in `RedisClient::Cluster` (or, at least, re-implement it in terms of `RedisClient::Cluster#with`.

My understanding was that the original idea of adding this kind of “pinning” was to avoid needing to implement transaction support in `RedisClient::Cluster` at all - instead users can issue transactions directly against the correct node selected by `#with`

Reply:
I'm wondering your decision. In Redis cluster, I think users should use the transaction feature as simple because CAP theory. I've thought the `#with` method is just an optional way to be able to use the transaction feature with some complecated ways in cluster mode. I've thought it's not the main interface for the transaction feature. It's a bit different from the redis-client. Please give me time to think it.

Reply:
Remember, the original problem I was solving was that `RedisClient::Cluster#multi` currently calls the transaction block twice - once itself to figure out what node to send the transaction to, and then `RedisClient#multi` calls it again to actually build up the transaction to execute. This interface is _also_ different from what's in `RedisClient`! 

From my pretty extensive survey of the options here over the last few months, there's a pretty fundamental tradeoff to make here:

* If you want to implement transaction support inside redis-cluster-client in terms of `RedisClient#multi`, you _must_ know the node ahead of time. You can't lazily call `RedisClient#multi` once you see the first command in the transaction, because you've already yielded control back to the caller, and can't give the caller the real `RedisClient::Multi` object without yielding again (which was the original bug I was trying to solve!).
* If you absolutely want `RedisClient::Cluster#multi` to work without knowing the node ahead of time (and thus, make it fully compatible), the only choice is to implement a kind of 'lazy transaction' system like I did in https://github.com/redis-rb/redis-cluster-client/pull/295. That is, you wait for the first command in the transaction, then issue MULTI on the appropriate node (and, thus, the redis-cluster-client must also take responsibility for calling EXEC/DISCARD/UNWATCH when appropriate too). This will work when issuing `client.call('MULTI') ... client.call('EXEC')`, which is valid too with `RedisClient`.

My understanding is that we settled on the `#with` interface as the best way to implement "knowing the node ahead of time", because you can also make the same code still work with RedisClient (since `RedisClient#with` does nothing - like I wrote in the docs here: https://github.com/redis-rb/redis-cluster-client/blob/86c0e01060a7f87f637f85d7fe4c64fe2255f70f/README.md?plain=1#L253-L255).

If we want to make `RedisClient::Cluster#multi` keep working how it is now, for backwards compatibility, we can keep its current implementation. However going forward I think we should be recommending people to use `RedisClient::Cluter#with` since it doesn't have the calls-block-twice behaviour.

For compatability with `RedisClient`, an approach like https://github.com/redis-rb/redis-cluster-client/pull/295 would be better - it would let us behave _exactly_ like a drop in replacement for `RedisClient`.